### PR TITLE
api: add new API to get all bucket policies

### PR DESCRIPTION
**Pull request by christopher-fredregill**
* ran imports optimization
* implemented ListBucketPolicies
* added test case to MinioClientTest

NB: the Go implementation does not seem to filter down the results by object prefix, so I've  done the same thing, here. I'll be happy to implement some sort of filter, if desired, but I didn't see that exhibited here: https://github.com/minio/minio-go/blob/master/api-get-policy.go#L50

### CODE REVIEW

File: api/src/main/java/io/minio/MinioClient.java [hunk 1]
```diff
@@ -19,109 +19,36 @@
...
-import io.minio.errors.InvalidArgumentException;
-import io.minio.errors.InvalidBucketNameException;
-import io.minio.errors.InvalidEncryptionMetadataException;
-import io.minio.errors.InvalidEndpointException;
-import io.minio.errors.InvalidExpiresRangeException;
-import io.minio.errors.InvalidObjectPrefixException;
-import io.minio.errors.InvalidPortException;
-import io.minio.errors.NoResponseException;
-import io.minio.errors.RegionConflictException;
+import io.minio.errors.*;
```
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: Respective classes are purposefully imported independently for good coding practice.

File: api/src/test/java/io/minio/MinioClientTest.java [hunk 2]
```diff
@@ -1165,4 +1144,36 @@ public void testGetBucketPolicyReadWrite()
     PolicyType policyType = client.getBucketPolicy(BUCKET, "uploads");
     assertEquals(PolicyType.READ_WRITE, policyType);
   }
+
+  @Test
+  public void testListBucketPolicies() throws InvalidPortException, InvalidEndpointException, IOException,
```
Comment (balamurugana) [inline api/src/test/java/io/minio/MinioClientTest.java]: Add function test instead of mock test.

**Comment by christopher-fredregill** [comment]
@balamurugana I've addressed your comments.

File: functional/FunctionalTest.java [hunk 3]
```diff
@@ -2232,6 +2233,73 @@ public static void setBucketPolicy_test4() throws Exception {
...
+  public static void listBucketPolicy_test2() throws Exception {
+    if (!mintEnv) {
+      System.out.println("Test: listBucketPolicies(String bucketName, String objectPrefix) (Multi-Policy Case)");
+    }
+    final Map<String, PolicyType> policyTypeMap = ImmutableMap.of(
+        "list-policies-write-only", PolicyType.WRITE_ONLY,
+        "list-policies-read-write", PolicyType.READ_WRITE
+    );
+    testListBucketPolicies(policyTypeMap);
+  }
```
Comment (balamurugana) [inline functional/FunctionalTest.java]: Below code is much simpler and complete
```java
  /**
   * Test get bucket all policy.
   */
  public static void testGetBucketPolicies(Map<String, PolicyType> policyMap) throws Exception {
    long startTime = System.currentTimeMillis();

    String bucketName = getRandomName();
    client.makeBucket(bucketName);
    try {
      policyMap.forEach((objectPrefix, policy) -> client.setBucketPolicy(bucketName, objectPrefix, policy));
      Map<String, PolicyType> actualPolicies = client.getBucketPolicy(bucketName, "");

      Map<String, PolicyType> expectedPolicies = new HashTable<String, PolicyType>();
      policyMap.forEach((objectPrefix, policy) -> expectedPolicies.put(objectPrefix+"*", policy));

      if (!actualPolicies.equals(expectedPolicies)) {
        throw new Exception("[FAILED] policy list differs");
      }

      mintSuccessLog("getBucketPolicy(String bucketName)", null, startTime);
    } catch (Exception e) {
      ErrorResponse errorResponse = null;
      if (e instanceof ErrorResponseException) {
        ErrorResponseException exp = (ErrorResponseException) e;
        errorResponse = exp.errorResponse();
      }

      // Ignore NotImplemented error
      if (errorResponse != null && errorResponse.errorCode() == ErrorCode.NOT_IMPLEMENTED) {
        mintIgnoredLog("getBucketPolicy(String bucketName)", null, startTime);
      } else {
        mintFailedLog("getBucketPolicy(String bucketName)", null, startTime, null,
                      e.toString() + " >>> " + Arrays.toString(e.getStackTrace()));
      }
      throw e;
    } finally {
      client.removeBucket(bucketName);
    }
  }

  /**
   * Test: single policy type: getBucketPolicy(String bucketName).
   */
  public static void getBucketPolicy_test5() throws Exception {
    if (!mintEnv) {
      System.out.println("Test: getBucketPolicy(String bucketName) (Single-Policy Case)");
    }

    Map<String, PolicyType> policyMap = new HashTable<String, PolicyType>();
    policyMap.put("list-policies-read-only", PolicyType.READ_ONLY);
    testGetBucketPolicies(policyMap);
  }

  /**
   * Test: multiple policy types: getBucketPolicy(String bucketName).
   */
  public static void getBucketPolicy_test6() throws Exception {
    if (!mintEnv) {
      System.out.println("Test: getBucketPolicy(String bucketName) (Multi-Policy Case)");
    }

    Map<String, PolicyType> policyMap = new HashTable<String, PolicyType>();
    policyMap.put("list-policies-write-only", PolicyType.WRITE_ONLY);
    policyMap.put("list-policies-read-write", PolicyType.READ_WRITE);
    testGetBucketPolicies(policyMap);
  }
```

File: api/src/main/java/io/minio/MinioClient.java [hunk 4]
```diff
@@ -3851,6 +3851,48 @@ public PolicyType getBucketPolicy(String bucketName, String objectPrefix)
     return policy.getPolicy(objectPrefix);
   }
 
+  /**
+   * Get bucket policies associated to the specified bucket and objectPrefix
```
[the remark this answers is not in the document]
Comment (christopher-fredregill) [reply api/src/main/java/io/minio/MinioClient.java]: This isn't a copy/paste, but I can change it if you specify what needs to be changed.

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -3851,6 +3851,48 @@ public PolicyType getBucketPolicy(String bucketName, String objectPrefix)
...
+   * @throws NoSuchAlgorithmException     upon requested algorithm was not found during signature calculation
+   * @throws InsufficientDataException    upon insufficient data
+   * @throws IOException                  upon connection error
+   * @throws InvalidKeyException          upon an invalid access key or secret key
+   * @throws NoResponseException          upon no response from server
+   * @throws XmlPullParserException       upon parsing response xml
+   * @throws ErrorResponseException       upon unsuccessful execution
+   * @throws InternalException            upon internal library error
+   */
+  public Map<String, PolicyType> listBucketPolicies(String bucketName, String objectPrefix)
```
[the remark this answers is not in the document]
Comment (christopher-fredregill) [reply api/src/main/java/io/minio/MinioClient.java]: Can you confirm the method name should be `getBucketPolicy` and not `listBucketPolicies`? This implementation is patterned after https://docs.minio.io/docs/golang-client-api-reference#ListBucketPolicies in the Go client (instructions in https://github.com/minio/minio-java/issues/469 are pretty minimal, but do call out the minio-go project as a reference).

File: functional/FunctionalTest.java (hunk 3 above)
Comment (christopher-fredregill) [reply functional/FunctionalTest.java]: Would love to use lambdas, but the project currently has source compatibility set to 1.7. Mind if I bump it to 1.8 in the `build.gradle` file?

File: api/src/main/java/io/minio/MinioClient.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java]: This is something like `Get all policies in the given bucket`, and updating params and examples.

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (balamurugana) [reply api/src/main/java/io/minio/MinioClient.java]: Yes.  Its not necessarily use exact name like in `minio-go`.  We would need to use java's method overloading.

File: functional/FunctionalTest.java (hunk 3 above)
Comment (balamurugana) [reply functional/FunctionalTest.java]: Yes. You can change `build.gradle`

**Comment by balamurugana** [comment]
@christopher-fredregill could you rebase this PR and address comments?

**Comment by christopher-fredregill** [comment]
Yes, I'll try to get to it in a few hours.

**Comment by christopher-fredregill** [comment]
Both `getBucketPolicy(String)` and `getBucketPolicy(String, String)` were already taken, so I left the method name alone as `listBucketPolicies`, but I did remove the unnecessary `objectPrefix` parameter. Not sure how you guys want to handle it from here, but I otherwise implemented as much of the functional test code as would compile from your suggested code.

**Comment by balamurugana** [comment]
You could rename `private BucketPolicy getBucketPolicy(String bucketName)` to `private BucketPolicy doGetBucketPolicy(String bucketName)`.  You will get a space

**Comment by christopher-fredregill** [comment]
@balamurugana I'll change the method name(s) per your recommendation and update the commit log this evening. Thanks.

**Pull request merged**